File: modulo04-agentes-autonomos/agent-evals-java/src/main/java/com/iadeva/evals/evaluator/BenchmarkRunner.java
package com.iadeva.evals.evaluator;

// Equivalente ao benchmark_runner.py da aula 09 — avalia comparativamente as 3 arquiteturas cognitivas

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.iadeva.evals.dataset.EvalDataset;
import com.iadeva.evals.model.BenchmarkMetrics;
import com.iadeva.evals.model.BenchmarkReport;
import com.iadeva.evals.model.EvalScenario;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.IOException;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

/**
 * Executa benchmarks comparativos entre as 3 arquiteturas cognitivas:
 * ReactAgent, PlanExecuteAgent e ReflectionAgent.
 * Os resultados são simulados com variações realistas para fins de avaliação do framework.
 */
@Component
public class BenchmarkRunner {

    private final EvalDataset evalDataset;
    private final ObjectMapper objectMapper;
    private final String reportsPath;

    // Semente fixa para reprodutibilidade dos resultados simulados
    private final Random random = new Random(42);

    public BenchmarkRunner(
            EvalDataset evalDataset,
            ObjectMapper objectMapper,
            @Value("${evals.reports.path:./reports}") String reportsPath
    ) {
        this.evalDataset = evalDataset;
        // Configura ObjectMapper com indentação e suporte a Instant
        this.objectMapper = objectMapper.copy()
                .enable(SerializationFeature.INDENT_OUTPUT)
                .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
                .findAndRegisterModules();
        this.reportsPath = reportsPath;
    }

    /**
     * Executa o benchmark completo e retorna o relatório comparativo.
     * O relatório também é salvo em disco em ./reports/.
     */
    public BenchmarkReport run() {
        List<EvalScenario> scenarios = evalDataset.getScenarios();

        // Simula execução dos 3 agentes em cada cenário
        BenchmarkMetrics react = simulateArchitecture("ReactAgent", scenarios, 0.82, 4.2, 1800);
        BenchmarkMetrics planExecute = simulateArchitecture("PlanExecuteAgent", scenarios, 0.88, 3.1, 2400);
        BenchmarkMetrics reflection = simulateArchitecture("ReflectionAgent", scenarios, 0.76, 5.8, 3200);

        List<BenchmarkMetrics> results = List.of(react, planExecute, reflection);

        // Gera veredicto comparativo por critério
        Map<String, String> verdict = gerarVeredicto(results);

        BenchmarkReport report = new BenchmarkReport(
                Instant.now(),
                scenarios.size(),
                results,
                verdict
        );

        salvarRelatorio(report);
        return report;
    }

    /**
     * Simula métricas realistas para uma arquitetura com pequenas variações aleatórias.
     *
     * @param architecture nome da arquitetura
     * @param scenarios    cenários do dataset
     * @param baseCompletion taxa base de conclusão
     * @param baseSteps      média base de steps
     * @param baseTokens     estimativa base de tokens
     */
    private BenchmarkMetrics simulateArchitecture(
            String architecture,
            List<EvalScenario> scenarios,
            double baseCompletion,
            double baseSteps,
            int baseTokens
    ) {
        // Aplica variação de ±5% sobre os valores base para simular ruído real
        double completionRate = clamp(baseCompletion + (random.nextDouble() - 0.5) * 0.10, 0, 1);
        double avgSteps = Math.max(1.0, baseSteps + (random.nextDouble() - 0.5) * 1.0);
        int avgTokens = (int) (baseTokens + (random.nextDouble() - 0.5) * 400);
        double toolSuccessRate = clamp(completionRate + 0.05 + random.nextDouble() * 0.05, 0, 1);

        // Calcula cobertura de ferramentas: percentual de expectedTools que seriam chamadas
        double totalExpected = scenarios.stream()
                .mapToInt(s -> s.expectedTools().size())
                .sum();
        double covered = totalExpected * clamp(completionRate + random.nextDouble() * 0.08, 0, 1);
        double toolCoverage = covered / totalExpected;

        return new BenchmarkMetrics(
                architecture,
                round(completionRate),
                round(avgSteps),
                avgTokens,
                round(toolSuccessRate),
                round(toolCoverage)
        );
    }

    /**
     * Gera veredicto identificando a arquitetura vencedora em cada critério.
     */
    private Map<String, String> gerarVeredicto(List<BenchmarkMetrics> results) {
        Map<String, String> verdict = new HashMap<>();

        // Melhor taxa de conclusão
        results.stream()
                .max((a, b) -> Double.compare(a.completionRate(), b.completionRate()))
                .ifPresent(m -> verdict.put("melhor_conclusao", m.architecture()));

        // Menor número de steps (mais eficiente)
        results.stream()
                .min((a, b) -> Double.compare(a.avgSteps(), b.avgSteps()))
                .ifPresent(m -> verdict.put("mais_eficiente", m.architecture()));

        // Menor consumo de tokens
        results.stream()
                .min((a, b) -> Integer.compare(a.avgTokensEstimated(), b.avgTokensEstimated()))
                .ifPresent(m -> verdict.put("menor_custo", m.architecture()));

        // Melhor cobertura de ferramentas
        results.stream()
                .max((a, b) -> Double.compare(a.toolCoverage(), b.toolCoverage()))
                .ifPresent(m -> verdict.put("melhor_cobertura_ferramentas", m.architecture()));

        // Recomendação geral: arquitetura com maior score ponderado
        String recomendado = results.stream()
                .max((a, b) -> Double.compare(scoreGeral(a), scoreGeral(b)))
                .map(BenchmarkMetrics::architecture)
                .orElse("indeterminado");
        verdict.put("recomendado_geral", recomendado);

        return verdict;
    }

    /**
     * Score ponderado para comparação geral: 40% conclusão + 30% cobertura + 30% toolSuccess.
     */
    private double scoreGeral(BenchmarkMetrics m) {
        return 0.40 * m.completionRate()
                + 0.30 * m.toolCoverage()
                + 0.30 * m.toolSuccessRate();
    }

    /**
     * Persiste o relatório em JSON no diretório configurado.
     */
    private void salvarRelatorio(BenchmarkReport report) {
        try {
            File dir = new File(reportsPath);
            if (!dir.exists()) {
                dir.mkdirs();
            }
            String timestamp = DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")
                    .withZone(java.time.ZoneId.systemDefault())
                    .format(report.generatedAt());
            File file = new File(dir, "benchmark_" + timestamp + ".json");
            objectMapper.writeValue(file, report);
        } catch (IOException e) {
            // Logar mas não falhar — o relatório em memória ainda é retornado
            System.err.println("[BenchmarkRunner] Falha ao salvar relatório: " + e.getMessage());
        }
    }

    private double clamp(double value, double min, double max) {
        return Math.max(min, Math.min(max, value));
    }

    private double round(double value) {
        return Math.round(value * 10000.0) / 10000.0;
    }
}

File: modulo04-agentes-autonomos/agent-evals-go/evaluator/benchmark_runner.go
// Package evaluator implementa os avaliadores do framework de evals —
// equivalente ao pacote evaluator/ da versão Java (aulas 09, 12, 15).
package evaluator

import (
	"encoding/json"
	"log"
	"math/rand"
	"os"
	"path/filepath"
	"time"

	"agent-evals/dataset"
	"agent-evals/model"
)

// BenchmarkRunner executa benchmarks comparativos simulados entre as 3
// arquiteturas cognitivas — equivalente a BenchmarkRunner.java (aula 09).
//
// Os resultados são simulados com variações realistas (seed fixo para
// reprodutibilidade) para fins de avaliação do framework, assim como na
// versão Java — a fonte de aleatoriedade é math/rand (não replica bit-a-bit
// a sequência de java.util.Random, mas preserva as mesmas faixas e
// distribuição estatística).
type BenchmarkRunner struct {
	Dataset     *dataset.EvalDataset
	ReportsPath string
	rng         *rand.Rand
}

// NewBenchmarkRunner cria um BenchmarkRunner com seed fixo (42).
func NewBenchmarkRunner(ds *dataset.EvalDataset, reportsPath string) *BenchmarkRunner {
	return &BenchmarkRunner{Dataset: ds, ReportsPath: reportsPath, rng: rand.New(rand.NewSource(42))}
}

// Run executa o benchmark completo, salva o relatório em disco e o retorna.
func (r *BenchmarkRunner) Run() model.BenchmarkReport {
	scenarios := r.Dataset.Scenarios()

	react := r.simulateArchitecture("ReactAgent", scenarios, 0.82, 4.2, 1800)
	planExecute := r.simulateArchitecture("PlanExecuteAgent", scenarios, 0.88, 3.1, 2400)
	reflection := r.simulateArchitecture("ReflectionAgent", scenarios, 0.76, 5.8, 3200)

	results := []model.BenchmarkMetrics{react, planExecute, reflection}

	report := model.BenchmarkReport{
		GeneratedAt:    time.Now(),
		TotalScenarios: len(scenarios),
		Results:        results,
		Verdict:        buildVerdict(results),
	}

	r.saveReport(report)
	return report
}

func (r *BenchmarkRunner) simulateArchitecture(architecture string, scenarios []model.EvalScenario, baseCompletion, baseSteps float64, baseTokens int) model.BenchmarkMetrics {
	completionRate := clamp(baseCompletion+(r.rng.Float64()-0.5)*0.10, 0, 1)
	avgSteps := max(1.0, baseSteps+(r.rng.Float64()-0.5)*1.0)
	avgTokens := int(float64(baseTokens) + (r.rng.Float64()-0.5)*400)
	toolSuccessRate := clamp(completionRate+0.05+r.rng.Float64()*0.05, 0, 1)

	totalExpected := 0
	for _, s := range scenarios {
		totalExpected += len(s.ExpectedTools)
	}

	var toolCoverage float64
	if totalExpected > 0 {
		covered := float64(totalExpected) * clamp(completionRate+r.rng.Float64()*0.08, 0, 1)
		toolCoverage = covered / float64(totalExpected)
	}

	return model.BenchmarkMetrics{
		Architecture:       architecture,
		CompletionRate:     round(completionRate),
		AvgSteps:           round(avgSteps),
		AvgTokensEstimated: avgTokens,
		ToolSuccessRate:    round(toolSuccessRate),
		ToolCoverage:       round(toolCoverage),
	}
}

// buildVerdict gera o veredicto identificando a arquitetura vencedora em
// cada critério.
func buildVerdict(results []model.BenchmarkMetrics) map[string]string {
	verdict := map[string]string{}

	best := results[0]
	for _, m := range results[1:] {
		if m.CompletionRate > best.CompletionRate {
			best = m
		}
	}
	verdict["melhor_conclusao"] = best.Architecture

	mostEfficient := results[0]
	for _, m := range results[1:] {
		if m.AvgSteps < mostEfficient.AvgSteps {
			mostEfficient = m
		}
	}
	verdict["mais_eficiente"] = mostEfficient.Architecture

	cheapest := results[0]
	for _, m := range results[1:] {
		if m.AvgTokensEstimated < cheapest.AvgTokensEstimated {
			cheapest = m
		}
	}
	verdict["menor_custo"] = cheapest.Architecture

	bestCoverage := results[0]
	for _, m := range results[1:] {
		if m.ToolCoverage > bestCoverage.ToolCoverage {
			bestCoverage = m
		}
	}
	verdict["melhor_cobertura_ferramentas"] = bestCoverage.Architecture

	recommended := results[0]
	for _, m := range results[1:] {
		if overallScore(m) > overallScore(recommended) {
			recommended = m
		}
	}
	verdict["recomendado_geral"] = recommended.Architecture

	return verdict
}

// overallScore é o score ponderado para comparação geral: 40% conclusão +
// 30% cobertura + 30% toolSuccess.
func overallScore(m model.BenchmarkMetrics) float64 {
	return 0.40*m.CompletionRate + 0.30*m.ToolCoverage + 0.30*m.ToolSuccessRate
}

// saveReport persiste o relatório em JSON no diretório configurado. Falhas
// são logadas mas não interrompem a execução — o relatório em memória ainda
// é retornado, igual à versão Java.
func (r *BenchmarkRunner) saveReport(report model.BenchmarkReport) {
	if err := os.MkdirAll(r.ReportsPath, 0o755); err != nil {
		log.Printf("[BenchmarkRunner] Falha ao salvar relatório: %v", err)
		return
	}

	timestamp := report.GeneratedAt.Format("20060102_150405")
	path := filepath.Join(r.ReportsPath, "benchmark_"+timestamp+".json")

	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		log.Printf("[BenchmarkRunner] Falha ao salvar relatório: %v", err)
		return
	}
	if err := os.WriteFile(path, data, 0o644); err != nil {
		log.Printf("[BenchmarkRunner] Falha ao salvar relatório: %v", err)
	}
}

func clamp(value, min, max float64) float64 {
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

func round(value float64) float64 {
	return float64(int64(value*10000+0.5)) / 10000
}
